File: Subsets-Backtrack.java
class Solution {
    
    List<List<Integer>> output;
    
    public List<List<Integer>> subsets(int[] nums) {
        
        output = new ArrayList<>();
        backtrack(nums,0,new ArrayList<>());
        
        return output;
        
    }
    
    private void backtrack(int[] nums, int start, List<Integer> currentPath){
        output.add(new ArrayList<>(currentPath));
        
        for (int x = start; x < nums.length; x++){
            currentPath.add(nums[x]);
            backtrack(nums,x+1,currentPath);
            currentPath.remove(currentPath.size() - 1);
        }
    }
}

File: subsetsBackTrack.py
class Solution:
    def __init__(self):
        self.result = []
    
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        if not nums:
            return None
        
        self.helper(nums, list(), 0)
        return self.result
    
    def helper(self, nums, path, i):
        #base
        if i==len(nums):
            self.result.append(list(path))
            return 
        #logic
        path.append(nums[i])
        self.helper(nums, path, i+1)
        path.pop(len(path)-1)
        self.helper(nums, path, i+1)
               
        
        
        
